actors/builtin/miner/miner_actor.go

```diff
@@ -487,6 +467,186 @@ func (a Actor) SubmitWindowedPoSt(rt Runtime, params *SubmitWindowedPoStParams)
+				active, err := partition.ActiveSectors()
+				if err != nil {
+					return err
+				}
+				err = faults.Add(partIdx, active)
+				if err != nil {
+					return err
+				}
+
+				// Record faulty power for penalties.
+				//
+				// NOTE: This also includes power that was
+				// activated at the end of the last challenge
+				// window, and power from sectors that have since
+				// expired. That means we may end up over-penalizing,
+				// but that's fine. The miner submitted a bad proof.
+				faultyPower = faultyPower.Add(partition.ActivePower())
+				return nil
+			})
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load partitions")
+
+			allSectorsNos, err := bitfield.MultiMerge(allSectors...)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to merge sector bitfields")
+
+			allIgnoredNos, err := bitfield.MultiMerge(allIgnored...)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to merge fault bitfields")
+
+			// Load sectors as seen at the end of the last proving period.
+			sectorsSnapshot, err := LoadSectors(store, dl.SectorsSnapshot)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load sectors array")
+			sectorInfos, err := sectorsSnapshot.LoadForProof(allSectorsNos, allIgnoredNos)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load sectors for post challenge")
+
+			// Find the proving period start for the deadline in question.
+			ppStart := st.ProvingPeriodStart
+			if st.CurrentDeadline < params.Deadline {
+				ppStart -= WPoStProvingPeriod
+			}
+
+			targetDeadline := NewDeadlineInfo(ppStart, params.Deadline, currEpoch)
+
+			// Fails if validation succeeds.
+			challengeWindowedPost(rt, targetDeadline.Challenge, sectorInfos, post.Proofs)
+
+			// Ok, so, here's the tricky part. Now we need to mark these all as _bad_. However, compaction could have
+			// happened, so the sectors could have moved.
+			//
+			// We've addressed this in two ways:
+			//
+			// 1. Compaction cannot be done for Finality epochs after a challenge window.
+			// 2. Successful challenges will always penalize, even if we can't find the target sectors and mark them faulty.
+			//    You can run but you can't hide.
+
+			// Load sectors. We can't use the snapshot as a sector
+			// with an extended expiration could end up being
+			// scheduled to expire at the wrong time.
+			//
+			// Unfortunately, this means we'll need to load all
+			// sector infos twice (once for the proof, once to mark
+			// faulty). We can't use memoized power info because
+			// sectors may have been terminated.
+			sectors, err := LoadSectors(store, st.Sectors)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load sectors array")
+			faultExpirationEpoch := targetDeadline.Last() + FaultMaxAge
+			powerDelta, err = dl.DeclareFaults(store, sectors, info.SectorSize, QuantSpecForDeadline(targetDeadline), faultExpirationEpoch, faults)
```

Comment:
`DeclareFaults` is an odd method to call here. I would expect something more like what happens in deadline cron when a PoSt is missed, which eventually calls `partition.RecordMissedPost()`.

If `DeclareFaults` really does the right thing for both this and ahead-of-time declaration, let's rename it to `RecordFaults` or similar.

Reply:
It does the same thing. `RecordMissedPost` doesn't call this because `RecordMissedPost` doesn't know which sectors are faulty (it just marks the entire partitions faulty).

But I agree on the rename.